### PR TITLE
Migrate last remaining state to RocksDB

### LINKED ISSUE #1188: Use RocksDB for workflow instance state

### LINKED ISSUE #1219: Workflow recovery on broker restart is no longer possible
In #1165 we removed the workflow fetching functionality and merged deployment processing into the workflow stream processor. Currently the deployed workflows are stored in so called `WorkflowCache`.
This cache is transient, which means that there is not snapshot made from the current state.
Since there are no snapshots of the workflows or deployments recovery is not possible on broker restart.

With #1188 we should fix this problem as well.

Note:

Restart without snapshots should work, since then reprocessing of the hole log is done.

### LINKED ISSUE #1355: Migrate Element instance index to rocks db
blocks #1188

### LINKED ISSUE #1384: RocksDB has to be opened with column families
Currently I cannot restart he broker which already contains state, i.e. a message was published. On restart the broker fails to open the state as it requires to know the column families before.

```
15:20:05.876 [service-container-main] [0.0.0.0:26501-zb-actors-0] ERROR io.zeebe.servicecontainer - Failed to build service                                                                             
org.rocksdb.RocksDBException: You have to open all column families. Column families not opened: subSendTime, msgSubscription, messageId, msgTimeToLive, keyColumn                                       
        at org.rocksdb.RocksDB.open(Native Method) ~[rocksdbjni-5.14.2.jar:?]
        at org.rocksdb.RocksDB.open(RocksDB.java:231) ~[rocksdbjni-5.14.2.jar:?]
        at io.zeebe.logstreams.state.StateController.openDB(StateController.java:111) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                              
        at io.zeebe.logstreams.state.StateController.open(StateController.java:83) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                 
        at io.zeebe.broker.util.KeyStateController.open(KeyStateController.java:53) ~[zeebe-broker-core-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                            
        at io.zeebe.broker.subscription.message.state.MessageStateController.open(MessageStateController.java:56) ~[zeebe-broker-core-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                              
        at io.zeebe.logstreams.state.StateSnapshotController.copySnapshot(StateSnapshotController.java:106) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                        
        at io.zeebe.logstreams.state.StateSnapshotController.recover(StateSnapshotController.java:79) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                              
        at io.zeebe.logstreams.processor.StreamProcessorController.recoverFromSnapshot(StreamProcessorController.java:154) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                         
        at io.zeebe.logstreams.processor.StreamProcessorController.onActorStarting(StreamProcessorController.java:129) ~[zb-logstreams-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                             
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:87) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                                          
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                                         
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:187) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                                      
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                       
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]                                                                                   
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zb-util-0.12.0-SNAPSHOT.jar:0.12.0-SNAPSHOT]
```

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* migrated the old ElementInstanceIndex to RocksDB
 * ElementInstance and IndexedRocks remains, the structure changed a bit
 * Three column families to store the instances
   * One to map the key to an instance
   * One to map the parent-child identifier to the instance (to find childs and parent nodes)
   * One to store records which are used later
 * Access to ElementInstances are cached, this means #getInstance
   will return every time the same object until flush is called

 * since the migration to rocks db is done with this PR recovery is now working again
 * fixed column family problem - they need to be defined on rocksDB opening

closes #1355
closes #1219
closes #1188 
closes #1384

### CODE REVIEW

**Review by ThorbenLindhauer — CHANGES_REQUESTED** [review]

**Comment by Zelldon** [comment]
@ThorbenLindhauer please take a look into the latest changes. I also did a bit of refactoring, for example  I saw that the PersistenceHelper can be used elsewhere.

File: broker-core/src/main/java/io/zeebe/broker/workflow/state/WorkflowState.java [hunk 1]
```diff
@@ -122,4 +141,53 @@ public boolean remove(WorkflowInstanceSubscriptionRecord record) {
...
+    return elementInstanceState.getStoredRecords(scopeKey);
+  }
+
+  public void consumeToken(long scopeKey) {
+    elementInstanceState.consumeToken(scopeKey);
+  }
+
+  public void spawnToken(long scopeKey) {
+    elementInstanceState.spawnToken(scopeKey);
+  }
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply broker-core/src/main/java/io/zeebe/broker/workflow/state/WorkflowState.java]: Sorry still need to do this

**Comment by ThorbenLindhauer** [comment]
Building this branch (with and without my last commit) produces quite a lot of failing test cases on my machine: https://gist.github.com/ThorbenLindhauer/915294b921513ad32edf02fd3322349b

File: broker-core/src/main/java/io/zeebe/broker/workflow/state/WorkflowState.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThorbenLindhauer) [reply broker-core/src/main/java/io/zeebe/broker/workflow/state/WorkflowState.java]: done, see https://github.com/zeebe-io/zeebe/pull/1389/commits/c14ce63be110396431b5aa1437d858ef94176526

**Comment by Zelldon** [comment]
Yes will fix the broken tests tomorrow thanks for the changes !

**Comment by Zelldon** [comment]
@ThorbenLindhauer Ok tests are running again :)  Please approve if everything is ok

**Review by ThorbenLindhauer — APPROVED** [review]

**Comment by Zelldon** [comment]
bors r=ThorbenLindhauer

**Pull request merged**